src/vs/workbench/contrib/chat/common/promptFileReference.ts

```diff
@@ -0,0 +1,400 @@
+
+/**
+ * Represents a file reference in the chatbot prompt, e.g. `#file:./path/to/file.md`.
+ * Contains logic to resolve all nested file references in the target file and all
+ * referenced child files recursively, if any.
+ *
+ * ## Examples
+ *
+ * ```typescript
+ * const fileReference = new PromptFileReference(
+ * 	 URI.file('/path/to/file.md'),
+ * 	 fileService,
+ * );
+ *
+ * // subscribe to updates to the file reference tree
+ * fileReference.onUpdate(() => {
+ * 	 // .. do something with the file reference tree ..
+ * 	 // e.g. get URIs of all resolved file references in the tree
+ * 	 const resolved = fileReference
+ * 		// get all file references as a flat array
+ * 		.flatten()
+ * 		// remove self from the list if only child references are needed
+ * 		.slice(1)
+ * 		// filter out unresolved references
+ * 		.filter(reference => reference.resolveFailed === flase)
+ * 		// convert to URIs only
+ * 		.map(reference => reference.uri);
+ *
+ * 	 console.log(resolved);
+ * });
+ *
+ * // *optional* if need to re-resolve file references when target files change
+ * // note that this does not sets up filesystem listeners for nested file references
+ * fileReference.addFilesystemListeners();
+ *
+ * // start resolving the file reference tree; this can also be `await`ed if needed
+ * // to wait for the resolution on the main file reference to complete (the nested
+ * // references can still be resolving in the background)
+ * fileReference.resolve();
+ *
+ * // don't forget to dispose when no longer needed!
+ * fileReference.dispose();
+ * ```
+ */
+export class PromptFileReference extends Disposable {
+	/**
+	 * Child references of the current one.
+	 */
+	protected readonly children: PromptFileReference[] = [];
+
+	private readonly _onUpdate = this._register(new Emitter<void>());
+	/**
+	 * The event is fired when nested prompt snippet references are updated, if any.
+	 */
+	public readonly onUpdate = this._onUpdate.event;
+
+	private _errorCondition?: TErrorCondition;
+	/**
+	 * If file reference resolution fails, this attribute will be set
+	 * to an error instance that describes the error condition.
+	 */
+	public get errorCondition(): TErrorCondition | undefined {
+		return this._errorCondition;
+	}
+
+	/**
+	 * Whether file reference resolution was attempted at least once.
+	 */
+	private _resolveAttempted: boolean = false;
+	/**
+	 * Whether file references resolution failed.
+	 * Set to `undefined` if the `resolve` method hasn't been ever called yet.
+	 */
+	public get resolveFailed(): boolean | undefined {
+		if (!this._resolveAttempted) {
+			return undefined;
+		}
+
+		return !!this._errorCondition;
+	}
+
+	constructor(
+		private readonly _uri: URI | Location,
+		@IFileService private readonly fileService: IFileService,
+		@IConfigurationService private readonly configService: IConfigurationService,
+	) {
+		super();
+		this.onFilesChanged = this.onFilesChanged.bind(this);
```

Comment:
What would be the "idiomatic" way for you in this case? Didn't want to rely on lexical scope binding since it wastes a lot of resources and makes the function context less flexible, nor wanted to re-create a new function often 🤔

Reply:
It's fine, and I don't mean to nit-pick, but it seems like a premature optimization where it's more readable to use a fat arrow function, which we do in many places. I prefer that because passing functions around with `bind` means it can be harder to track where the listener was attached when debugging, and because I just find 'bind' to be ugly.

Reply:
Well, it seems to be a subjective code style preference - for me passing the `this.callback` to a listener is shorter and more readable way 🤷 Talking about ideomatic way of binding context - this is precisely it, as the arrow function does not really bind the context, and I know folks that think that the arrow functions are ugly and in most of the cases lead to not inflexible and inefficient code 🤔

These things should be defined in linter config or at least in codestyle guides, otherwise everybody uses whatever they see fit ([one](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/chat/browser/chatWidget.ts#L669) [two](https://github.com/microsoft/vscode/blob/main/extensions/git/src/ipc/ipcServer.ts#L75) [three](https://github.com/microsoft/vscode/blob/f408139c28c2c653d673e9a85ba250d6c6a588be/extensions/vscode-colorize-perf-tests/test/colorize-fixtures/test-treeView.ts#L690) etc.).